File: bench/testdata/samples/A10/PSB-A10-PREV-001/javascript/secure.js
// PSB-A10-PREV-001: URL preview/unfurl feature
// CWE: CWE-918
// Expected: (none - secure)
const express = require("express");
const axios = require("axios");
const cheerio = require("cheerio");
const dns = require("dns").promises;
const { URL } = require("url");
const net = require("net");
const app = express();
app.use(express.json());

async function isSafeUrl(urlStr) {
  const parsed = new URL(urlStr);
  if (!["http:", "https:"].includes(parsed.protocol)) return false;
  const addresses = await dns.resolve4(parsed.hostname);
  for (const addr of addresses) {
    if (net.isIP(addr) && (addr.startsWith("10.") || addr.startsWith("127.") ||
        addr.startsWith("169.254.") || addr.startsWith("192.168.") ||
        addr.startsWith("172.16.") || addr === "0.0.0.0")) {
      return false;
    }
  }
  return true;
}

app.post("/api/preview", async (req, res) => {
  const { url } = req.body;
  if (!url) {
    return res.status(400).json({ error: "url is required" });
  }

  try {
    if (!(await isSafeUrl(url))) {
      return res.status(400).json({ error: "blocked URL" });
    }

    const resp = await axios.get(url, {
      timeout: 5000,
      maxRedirects: 0,
      maxContentLength: 1024 * 1024,
    });
    const $ = cheerio.load(resp.data);

    const title = $("title").text() || "";
    const description = $('meta[name="description"]').attr("content") || "";
    const image = $('meta[property="og:image"]').attr("content") || "";

    res.json({ title, description, image });
  } catch (err) {
    res.status(400).json({ error: "failed to fetch URL" });
  }
});

module.exports = app;

File: bench/testdata/samples/A10/PSB-A10-PREV-001/python/secure.py
# PSB-A10-PREV-001: URL preview/unfurl feature
# CWE: CWE-918
# Expected: (none - secure)
from flask import Flask, request, jsonify
import requests
import ipaddress
import socket
from urllib.parse import urlparse
from bs4 import BeautifulSoup

app = Flask(__name__)

BLOCKED_PORTS = {22, 25, 3306, 5432, 6379}

def is_safe_url(url):
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https"):
        return False
    hostname = parsed.hostname
    if not hostname:
        return False
    if parsed.port and parsed.port in BLOCKED_PORTS:
        return False
    try:
        addr = socket.getaddrinfo(hostname, None)[0][4][0]
        ip = ipaddress.ip_address(addr)
        if ip.is_private or ip.is_loopback or ip.is_reserved:
            return False
    except (socket.gaierror, ValueError):
        return False
    return True

@app.route("/api/preview", methods=["POST"])
def preview():
    data = request.get_json()
    url = data.get("url")
    if not url:
        return jsonify({"error": "url is required"}), 400
    if not is_safe_url(url):
        return jsonify({"error": "blocked URL"}), 400

    try:
        resp = requests.get(url, timeout=5, allow_redirects=False)
        soup = BeautifulSoup(resp.text, "html.parser")

        title = soup.title.string if soup.title else ""
        desc_tag = soup.find("meta", attrs={"name": "description"})
        description = desc_tag["content"] if desc_tag else ""
        img_tag = soup.find("meta", attrs={"property": "og:image"})
        image = img_tag["content"] if img_tag else ""

        return jsonify({"title": title, "description": description, "image": image})
    except Exception:
        return jsonify({"error": "failed to fetch URL"}), 400

if __name__ == "__main__":
    app.run()
